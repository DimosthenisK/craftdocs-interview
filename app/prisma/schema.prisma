// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
    provider = "prisma-client-js"
}

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

model Document {
    id String @unique @default(nanoid())

    content String

    owner   User   @relation(fields: [ownerId], references: [id], onDelete: Cascade, onUpdate: Cascade)
    ownerId String

    createdAt DateTime @default(now())
    updatedAt DateTime @default(now()) @updatedAt

    subscriptions DocumentSubscription[]
}

model DocumentSubscription {
    id String @unique @default(nanoid())

    lastSync DateTime @default(now())

    document   Document @relation(fields: [documentId], references: [id], onDelete: Cascade, onUpdate: Cascade)
    documentId String
    user       User     @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
    userId     String

    createdAt DateTime @default(now())
    updatedAt DateTime @default(now()) @updatedAt

    @@unique([documentId, userId])
}

model User {
    id String @unique @default(nanoid())

    name String @unique // To support anonymous users

    createdAt DateTime @default(now())
    updatedAt DateTime @default(now()) @updatedAt

    subscriptions  DocumentSubscription[]
    ownedDocuments Document[]
}
