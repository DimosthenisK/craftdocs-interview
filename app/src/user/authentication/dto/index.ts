export * from './check-token.dto';
export * from './login.dto';
export * from './responses';
