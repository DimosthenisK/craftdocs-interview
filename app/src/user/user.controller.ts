import { Body, Controller, Get, Patch, UseGuards } from '@nestjs/common';
import { ApiBearerAuth, ApiOkResponse, ApiTags } from '@nestjs/swagger';
import { User } from '@prisma/client';
import { BearerAuthGuard } from './authentication/guards/bearer.guard';
import { CurrentUser } from './decorators';
import { UpdateUserDto } from './dto/update-user.dto';
import { UserEntity } from './entities/user.entity';
import { UserService } from './user.service';

@ApiTags('User')
@ApiBearerAuth('bearer')
@UseGuards(BearerAuthGuard)
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get()
  @ApiOkResponse({ type: UserEntity })
  findOne(@CurrentUser() user: User) {
    return this.userService.findOne(user.id);
  }

  @Patch()
  @ApiOkResponse({ type: UserEntity })
  update(@CurrentUser() user: User, @Body() updateUserDto: UpdateUserDto) {
    return this.userService.update(user.id, updateUserDto);
  }
}
