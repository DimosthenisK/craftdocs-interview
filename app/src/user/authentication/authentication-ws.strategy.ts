import { ExtractJwt, Strategy } from 'passport-jwt';

import { ConfigService } from '@nestjs/config';
import { Injectable } from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { User } from '@prisma/client';
import { UserService } from '../user.service';
import { WsException } from '@nestjs/websockets';
import { tokenInfo } from './authentication.service';

@Injectable()
export class WsJwtStrategy extends PassportStrategy(Strategy, 'wsjwt') {
  constructor(
    private readonly userService: UserService,
    private readonly configService: ConfigService,
  ) {
    super({
      jwtFromRequest: ExtractJwt.fromUrlQueryParameter('bearerToken'),
      secretOrKey: configService.get<string>('JWT_SECRET'),
    });
  }

  async validate({ id, name }: tokenInfo): Promise<User> {
    try {
      return this.userService.findOneByName(name);
    } catch (error) {
      throw new WsException('Unauthorized access');
    }
  }
}
