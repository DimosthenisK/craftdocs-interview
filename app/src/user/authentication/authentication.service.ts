import * as jwt from 'jsonwebtoken';

import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { User } from '@prisma/client';
import { UserService } from '../user.service';

export interface tokenInfo {
  id: string;
  name: string;
}

@Injectable()
export class AuthenticationService {
  constructor(
    private readonly usersService: UserService,
    private readonly configService: ConfigService,
  ) {}

  async validateUser(token: string): Promise<User> {
    let tokenInfo;
    try {
      tokenInfo = jwt.verify(
        token,
        this.configService.get('JWT_SECRET'),
      ) as tokenInfo;
    } catch (err) {
      throw err;
    }
    return await this.usersService.findOne(tokenInfo.id);
  }

  async generateTokenForUser(user: User, expiration?: string): Promise<string> {
    expiration = expiration
      ? expiration
      : this.configService.get('DEFAULT_TOKEN_EXPIRATION_INTERVAL') || '7d';
    return jwt.sign(
      {
        id: user.id,
        name: user.name,
      },
      this.configService.get('JWT_SECRET'),
      { expiresIn: expiration },
    );
  }

  /**
   * Under a production environment, this method should be modified to
   * support password authentication.
   * Here, we just check if a user with $name exists, and if he does, we log him in.
   * If not, we create him and log him in.
   */
  async login(name: string) {
    let user = await this.usersService.findOneByName(name);
    if (!user) {
      user = await this.usersService.create({ name });
    }

    return this.generateTokenForUser(user);
  }

  async checkToken(token: string) {
    let user;
    try {
      user = await this.validateUser(token);
    } catch (err) {
      if (err instanceof jwt.TokenExpiredError) {
        throw new Error('Token has expired');
      } else if (err instanceof jwt.JsonWebTokenError) {
        throw new Error('Token signature seems invalid or otherwise modified');
      } else {
        throw new Error('An unknown error has occured.');
      }
    }

    if (user) return this.generateTokenForUser(user);
    else throw new Error('User not found');
  }
}
