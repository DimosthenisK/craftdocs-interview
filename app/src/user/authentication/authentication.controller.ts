import {
  BadRequestException,
  Body,
  Controller,
  HttpCode,
  Post,
} from '@nestjs/common';
import { ApiOkResponse, ApiTags } from '@nestjs/swagger';
import { AuthenticationService } from './authentication.service';
import { CheckTokenDto, LoginDto, TokenResponse } from './dto';

@ApiTags('User Authentication')
@Controller('authentication')
export class AuthenticationController {
  constructor(public authenticationService: AuthenticationService) {}

  @Post('/login')
  @HttpCode(200)
  @ApiOkResponse({ type: () => TokenResponse })
  async login(@Body() body: LoginDto): Promise<TokenResponse> {
    const token = await this.authenticationService.login(body.name);

    // In a production environment
    // login() wouldn't always return a token.
    // See checkToken on how we'd handle that.

    return {
      token,
    };
  }

  /**
   * This flow isn't really needed here, as we could just always call login() in this implementation.
   * It's a good example of how to a true login flow would work.
   */
  @Post('/check-token')
  @HttpCode(200)
  @ApiOkResponse({ type: () => TokenResponse })
  async checkToken(@Body() body: CheckTokenDto): Promise<TokenResponse> {
    try {
      const token = await this.authenticationService.checkToken(body.token);

      return {
        token,
      };
    } catch (err) {
      throw new BadRequestException(err.message);
    }
  }
}
