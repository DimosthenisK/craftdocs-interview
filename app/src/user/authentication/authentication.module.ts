import { Module, forwardRef } from '@nestjs/common';

import { AuthenticationController } from './authentication.controller';
import { AuthenticationService } from './authentication.service';
import { BearerAuthGuard } from './guards/bearer.guard';
import { HttpStrategy } from './authentication.strategy';
import { PassportModule } from '@nestjs/passport';
import { UserModule } from '../user.module';
import { WsAuthGuard } from './guards/ws.guard';
import { WsJwtStrategy } from './authentication-ws.strategy';

@Module({
  imports: [PassportModule, forwardRef(() => UserModule)],
  providers: [
    AuthenticationService,
    HttpStrategy,
    BearerAuthGuard,
    WsJwtStrategy,
    WsAuthGuard,
  ],
  controllers: [AuthenticationController],
  exports: [
    AuthenticationService,
    PassportModule,
    BearerAuthGuard,
    WsAuthGuard,
  ],
})
export class AuthenticationModule {}
