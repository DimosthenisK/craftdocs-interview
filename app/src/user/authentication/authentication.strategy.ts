import { Injectable, UnauthorizedException } from '@nestjs/common';

import { AuthenticationService } from './authentication.service';
import { PassportStrategy } from '@nestjs/passport';
import { Strategy } from 'passport-http-bearer';

@Injectable()
export class HttpStrategy extends PassportStrategy(
  Strategy,
  'user-authentication',
) {
  constructor(private readonly authService: AuthenticationService) {
    super();
  }

  async validate(token: string) {
    try {
      const user = await this.authService.validateUser(token);
      if (!user) {
        throw new Error('User not found');
      }
      return user;
    } catch (err) {
      throw new UnauthorizedException(err);
    }
  }
}
