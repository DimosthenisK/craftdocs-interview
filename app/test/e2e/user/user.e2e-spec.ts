import * as request from 'supertest';

import { AuthenticationService } from '../../../src/user/authentication/authentication.service';
import { CacheModule } from '@nestjs/cache-manager';
import { ConfigModule } from '@nestjs/config';
import { INestApplication } from '@nestjs/common';
import { PrismaModule } from '../../../src/prisma/prisma.module';
import { PrismaService } from '../../../src/prisma/prisma.service';
import { Test } from '@nestjs/testing';
import { UserModule } from '../../../src/user/user.module';
import { UserService } from '../../../src/user/user.service';
import { buildTest } from '../helpers';

describe('UserController (e2e)', () => {
  jest.setTimeout(600000);
  const testID = `user-container-test`;

  let app: INestApplication;
  let prismaService: PrismaService;
  let authService: AuthenticationService;
  let userService: UserService;

  let afterAllFn: () => Promise<void>;
  let beforeEachFn: () => Promise<void>;
  let afterEachFn: () => Promise<void>;

  beforeAll(async () => {
    const moduleFixture = Test.createTestingModule({
      imports: [
        PrismaModule,
        ConfigModule.forRoot({
          isGlobal: true,
        }),
        CacheModule.register({ isGlobal: true }),
        UserModule,
      ],
    });

    const test = await buildTest(testID, moduleFixture);

    app = test.app;
    prismaService = test.prismaService;
    authService = test.authService;
    userService = app.get(UserService);

    afterAllFn = test.afterAll;
    beforeEachFn = test.beforeEach;
    afterEachFn = test.afterEach;
  });

  afterAll(async () => {
    await afterAllFn();
  });

  beforeEach(async () => {
    await beforeEachFn();
  });

  afterEach(async () => {
    await afterEachFn();
  });

  describe('GET /user', () => {
    it('should return the currently logged in user', async () => {
      const mockUser = await userService.create({ name: 'Mock User' });
      const mockUserToken = await authService.generateTokenForUser(mockUser);

      const response = await request(app.getHttpServer())
        .get(`/user`)
        .set('Authorization', `Bearer ${mockUserToken}`)
        .expect(200);

      expect(response.body.id).toMatch(mockUser.id);
      expect(response.body.name).toMatch(mockUser.name);
    });
  });

  describe('PATCH /user', () => {
    it('should update the currently logged in user', async () => {
      const mockUser = await userService.create({ name: 'Mock User' });
      const mockUserToken = await authService.generateTokenForUser(mockUser);

      const response = await request(app.getHttpServer())
        .patch(`/user`)
        .set('Authorization', `Bearer ${mockUserToken}`)
        .send({ name: 'New Name' })
        .expect(200);

      expect(response.body.id).toMatch(mockUser.id);
      expect(response.body.name).toMatch('New Name');
    });

    it('should not update the currently logged in user if the name is too long', async () => {
      const mockUser = await userService.create({ name: 'Mock User' });
      const mockUserToken = await authService.generateTokenForUser(mockUser);

      const response = await request(app.getHttpServer())
        .patch(`/user`)
        .set('Authorization', `Bearer ${mockUserToken}`)
        .send({ name: 'New Name'.repeat(100) })
        .expect(400);

      expect(response.body.errorCode).toMatchSnapshot();
    });

    it('should not work if the user isnt logged in', async () => {
      const mockUser = await userService.create({ name: 'Mock User' });
      const mockUserToken = await authService.generateTokenForUser(mockUser);

      const response = await request(app.getHttpServer())
        .patch(`/user`)
        //.set('Authorization', `Bearer ${mockUserToken}`)
        .send({ name: 'New Name'.repeat(100) })
        .expect(401);

      expect(response.body.errorCode).toMatchSnapshot();
    });
  });
});
