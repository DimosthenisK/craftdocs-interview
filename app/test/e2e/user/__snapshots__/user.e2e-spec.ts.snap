// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController (e2e) PATCH /user should not update the currently logged in user if the name is too long 1`] = `"VALIDATION_EXCEPTION"`;

exports[`UserController (e2e) PATCH /user should not work if the user isnt logged in 1`] = `undefined`;
